posthog-ai/src/langchain/callbacks.ts

```diff
@@ -0,0 +1,578 @@
+    const traceId = this._getTraceId(runId)
+    this._popParentOfRun(runId)
+    const run = this._popRunMetadata(runId)
+    if (!run || !('modelParams' in run)) {
+      console.warn(`Run ${runId} is not a generation, but attempted to be captured as such.`)
+      return
+    }
+    const actualParentRunId = this._getParentRunId(traceId, runId, parentRunId)
+    this._captureGeneration(traceId, runId, run as GenerationMetadata, response, actualParentRunId)
+  }
+
+  private _captureGeneration(
+    traceId: string,
+    runId: string,
+    run: GenerationMetadata,
+    output: LLMResult | Error,
+    parentRunId?: string
+  ): void {
+    const latency = run.endTime ? (run.endTime - run.startTime) / 1000 : 0
+    const eventProperties: Record<string, any> = {
+      $ai_trace_id: traceId,
+      $ai_span_id: runId,
+      $ai_span_name: run.name,
+      $ai_parent_id: parentRunId,
+      $ai_provider: run.provider,
+      $ai_model: run.model,
+      $ai_model_parameters: run.modelParams,
+      $ai_input: withPrivacyMode(this.client, this.privacyMode, run.input),
+      $ai_http_status: 200,
+      $ai_latency: latency,
+      $ai_base_url: run.baseUrl,
+    }
+
+    if (output instanceof Error) {
+      eventProperties['$ai_http_status'] = (output as any).status || 500
+      eventProperties['$ai_error'] = output.toString()
+      eventProperties['$ai_is_error'] = true
+    } else {
+      // Handle token usage
+      const [inputTokens, outputTokens] = this.parseUsage(output)
+      eventProperties['$ai_input_tokens'] = inputTokens
+      eventProperties['$ai_output_tokens'] = outputTokens
+
+      // Handle generations/completions
+      let completions
+      if (output.generations && Array.isArray(output.generations)) {
+        const lastGeneration = output.generations[output.generations.length - 1]
+        if (Array.isArray(lastGeneration)) {
+          completions = lastGeneration.map((gen) => {
+            return { role: 'assistant', content: gen.text }
+          })
+        }
+      }
+
+      if (completions) {
+        eventProperties['$ai_output_choices'] = withPrivacyMode(this.client, this.privacyMode, completions)
+      }
+    }
+
+    Object.assign(eventProperties, this.properties)
+    if (!this.distinctId) {
+      eventProperties['$process_person_profile'] = false
+    }
+
+    this.client.capture({
+      distinctId: (this.distinctId as string) || traceId,
+      event: '$ai_generation',
+      properties: eventProperties,
+      groups: this.groups,
+    })
+  }
+
+  private _logDebugEvent(eventName: string, runId: string, parentRunId: string | undefined, extra: any): void {
+    if (this.debug) {
+      console.log(`Event: ${eventName}, runId: ${runId}, parentRunId: ${parentRunId}, extra:`, extra)
+    }
+  }
+
+  private _getLangchainRunName(serialized: any, ...args: any[]): string | undefined {
+    if (args && args.length > 0) {
+      for (const arg of args) {
+        if (arg && arg.name) return arg.name
```

Comment:
It might be better actually to check if the arg is an object like:
```suggestion
        if (arg && typeof arg === 'object' && 'name' in arg && typeof arg.name === 'string') {
          return arg.name
        }
```
It might be an overkill, though.